src/ping.c

```diff
@@ -431,151 +486,188 @@ static int ping_init(void) /* {{{ */
   return start_thread();
 } /* }}} int ping_init */
 
-static int config_set_string(const char *name, /* {{{ */
-                             char **var, const char *value) {
-  char *tmp;
 
-  tmp = strdup(value);
-  if (tmp == NULL) {
-    ERROR("ping plugin: Setting `%s' to `%s' failed: strdup failed: %s", name,
-          value, STRERRNO);
+static int ping_config_add (oconfig_item_t *ci, bool is_device) {
+  sourcelist_t *sl;
+  sl = calloc(1, sizeof(*sl));
+
+  if (sl == NULL) {
+    ERROR("ping plugin: calloc failed in ping_config_add.");
+    return 1;
+  }
+  char *source = NULL;
+  if (cf_util_get_string(ci, &source) != 0) {
+    sfree(sl);
+    ERROR("ping plugin: strdup failed: %s in ping_config_add", STRERRNO);
     return 1;
   }
+  sl->source = source;
+  sl->is_device = is_device;
+  sl->next = sourcelist_head;
+  sl->ping_af = PING_DEF_AF;
+  sourcelist_head = sl;
+  sl->hosts = NULL;
+
+  for (int i = 0; i < ci->children_num; i++) {
+    oconfig_item_t *child = ci->children + i;
+    if (strcasecmp(child->key, "Host") == 0) {
+      hostlist_t *hl;
+      char *host = NULL;
+      hl = malloc(sizeof(*hl));
```

Comment:
Please use `calloc` instead of `malloc` so that the buffer is zeroed.

```suggestion
      hostlist_t *hl = calloc(1, sizeof(*hl));
      /* declare `host` just before calling `cf_util_get_string` */
```